cpp/unreal_plugins/SimulationController/Source/SimulationController/VehicleAgent.cpp

```diff
@@ -0,0 +1,266 @@
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 4 };
+        observation_space["wheel_encoder"] = std::move(array_desc); // FL, FR, RL, RR, in [rad/s]
+    }
+
+    if (Std::contains(observation_components, "imu")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 6 };
+        observation_space["imu"] = std::move(array_desc); // a_x, a_y, a_z in [cm/s^s] g_x, g_y, g_z in [rad/s]
+    }
+
+    std::map<std::string, ArrayDesc> camera_sensor_observation_space = camera_sensor_->getObservationSpace(observation_components);
+    for (auto& camera_sensor_observation_space_component : camera_sensor_observation_space) {
+        observation_space[camera_sensor_observation_space_component.first] = std::move(camera_sensor_observation_space_component.second);
+    }
+
+    return observation_space;
+}
+
+std::map<std::string, ArrayDesc> VehicleAgent::getStepInfoSpace() const
+{
+    return {};
+}
+
+void VehicleAgent::applyAction(const std::map<std::string, std::vector<uint8_t>>& action)
+{
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.VEHICLE_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "set_drive_torques")) {
+        vehicle_pawn_->setDriveTorques(Std::reinterpretAs<double>(action.at("set_drive_torques")));
+        vehicle_pawn_->setBrakeTorques(std::vector<double>{0.0, 0.0, 0.0, 0.0});
+    } else if (Std::contains(action_components, "set_brake_torques")) {
+        vehicle_pawn_->setBrakeTorques(Std::reinterpretAs<double>(action.at("set_brake_torques")));
+    }
+}
+
+std::map<std::string, std::vector<uint8_t>> VehicleAgent::getObservation() const
+{
+    std::map<std::string, std::vector<uint8_t>> observation;
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "position")) {
+        FVector location = vehicle_pawn_->GetActorLocation();
+        observation["position"] = Std::reinterpretAs<uint8_t>(std::vector<double>{
+            location.X,
+            location.Y,
+            location.Z});
+    }
+
+    if (Std::contains(observation_components, "orientation")) {
+        FRotator rotation = vehicle_pawn_->GetActorRotation();
+        observation["orientation"] = Std::reinterpretAs<uint8_t>(std::vector<double>{
+            rotation.Pitch,
+            rotation.Yaw,
+            rotation.Roll});
+    }
+
+    if (Std::contains(observation_components, "wheel_encoder")) {
+        observation["wheel_encoder"] = Std::reinterpretAs<uint8_t>(vehicle_pawn_->getWheelRotationSpeeds());
```

Comment:
`vehicle_pawn_->vehicle_movement_component_->getWheelRotationSpeeds()` to avoid trivial pass-through layers.